### ISSUE TITLE
NRE in Diagnostics UpdateMemberRange

### ISSUE THREAD

**Issue by balajikris** [dotnet/roslyn#5894]
With Roslyn bits 20151011.2, VS immediately crashes on typing code, with this stack.
~~~
Application: devenv.exe
Framework Version: v4.0.30319
Description: The application requested process termination through System.Environment.FailFast(string message).
Message: System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.CodeAnalysis.Text.TextSpan..ctor(Int32 start, Int32 length)
   at Microsoft.CodeAnalysis.Diagnostics.EngineV1.MemberRangeMap.UpdateMemberRange_NoLock(DictionaryData data, Document document, VersionStamp newVersion, Int32 memberId, TextSpan currentSpan, VersionStamp oldVersion)
   at Microsoft.CodeAnalysis.Diagnostics.EngineV1.MemberRangeMap.UpdateMemberRange(DiagnosticAnalyzer analyzer, Document document, VersionStamp newVersion, Int32 memberId, TextSpan span, MemberRanges oldRanges)
   at Microsoft.CodeAnalysis.Diagnostics.EngineV1.DiagnosticIncrementalAnalyzer.<AnalyzeBodyDocumentAsync>d__52.MoveNext()
Stack:
   at System.Environment.FailFast(System.String, System.Exception)
   at Microsoft.CodeAnalysis.FailFast.OnFatalException(System.Exception)
   at Microsoft.CodeAnalysis.ErrorReporting.FatalError.Report(System.Exception, System.Action`1<System.Exception>)
   at Microsoft.CodeAnalysis.ErrorReporting.FatalError.ReportUnlessCanceled(System.Exception)
   at Microsoft.CodeAnalysis.Diagnostics.EngineV1.DiagnosticIncrementalAnalyzer+<AnalyzeBodyDocumentAsync>d__52.MoveNext()
   at Microsoft.CodeAnalysis.Text.TextSpan..ctor(Int32, Int32)
   at Microsoft.CodeAnalysis.Diagnostics.EngineV1.MemberRangeMap.UpdateMemberRange_NoLock(DictionaryData, Microsoft.CodeAnalysis.Document, Microsoft.CodeAnalysis.VersionStamp, Int32, Microsoft.CodeAnalysis.Text.TextSpan, Microsoft.CodeAnalysis.VersionStamp)
   at Microsoft.CodeAnalysis.Diagnostics.EngineV1.MemberRangeMap.UpdateMemberRange(Microsoft.CodeAnalysis.Diagnostics.DiagnosticAnalyzer, Microsoft.CodeAnalysis.Document, Microsoft.CodeAnalysis.VersionStamp, Int32, Microsoft.CodeAnalysis.Text.TextSpan, MemberRanges)
   at Microsoft.CodeAnalysis.Diagnostics.EngineV1.DiagnosticIncrementalAnalyzer+<AnalyzeBodyDocumentAsync>d__52.MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.InvokeMoveNext(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.Run()
   at System.Threading.Tasks.AwaitTaskContinuation.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()
~~~

**Comment by balajikris** [comment]
@mavasani is taking a look, thanks Manish!

**Comment by mavasani** [comment]
I am able to repro this after InstallDogfood but unable to repro it from the RoslynDev hive.. the exception itself seems weird as there is nothing in the [TextSpan ctor](http://source.roslyn.io/Microsoft.CodeAnalysis/R/e229722e0f1b3dd9.html) that can throw an NRE
```
Message: System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.CodeAnalysis.Text.TextSpan..ctor(Int32 start, Int32 length)
```

Investigating..

**Comment by AdamSpeight2008** [comment]
Maybe it is in the `DiagnosticIncrementalAnalyzer` ?

**Comment by mavasani** [comment]
Okay, so I verified this is indeed a compiler bug. PR https://github.com/dotnet/roslyn/pull/5848 moved Roslyn.sln to a newer compiler toolset and the crash is being caused by corrupt codegen for Microsoft.CodeAnalysis.Features.dll. I locally reverted the commit https://github.com/dotnet/roslyn/commit/3e4cabdcf9d56e073e47f8c2176abc6f97cd81fc for that PR, build Microsoft.CodeAnalysis.Features.dll in release configuration with OfficialBuild = true and patched the 14.0 hive with the locally built features assembly and I see no crash.

**Comment by mavasani** [comment]
In fact, even our vsupdate patch from vsuml branch is completely busted and hits this crash.

@jmarolf Can you please revert your PR #5848 as soon as possible?

**Comment by mavasani** [comment]
FYI @tannergooding

**Comment by jmarolf** [comment]
I've reverted the toolset update. Can we get someone from the compiler team to investigate this?

**Comment by jaredpar** [comment]
This stack seems broken ... how can we possibly have a NRE in the TextSpan ctor?  It's a struct that sets fields of type int.  

@VSadov  can you take a look at this?

**Comment by pharring** [comment]
This is the same thing we've been hitting in the perf lab. See #5880

**Comment by VSadov** [comment]
This is because of #4828 which have been fixed. 

There were two bugs in the compiler one indirectly preventing another from reproing in some scenarios. You have managed to pick the compiler at the point when the first bug was fixed and introduced more cases for the second bug.

The binaries need to be rebuilt with newer compiler where #4828 is fixed.

**Comment by jaredpar** [comment]
Closing as this is already fixed in #4828
